File: app/src/main/java/com/sardox/timestamper/dialogs/AddCategoryDialog.java
package com.sardox.timestamper.dialogs;

import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;

import com.sardox.timestamper.R;
import com.sardox.timestamper.objects.Category;
import com.sardox.timestamper.recyclerview.IconAdapter;
import com.sardox.timestamper.types.JetUUID;
import com.sardox.timestamper.utils.Consumer;
import com.sardox.timestamper.utils.TimestampIcon;

import java.util.List;

public class AddCategoryDialog {

    public AddCategoryDialog(Context context, List<TimestampIcon> icons, final Consumer<Category> onCategoryCreated) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        View viewInflated = LayoutInflater.from(context).inflate(R.layout.new_category, null, false);
        final EditText input = (EditText) viewInflated.findViewById(R.id.input_cat);
        final IconAdapter iconPicker = new IconAdapter(icons, new Consumer<TimestampIcon>() {
            @Override
            public void accept(TimestampIcon icon) {
                input.setText(icon.getDescription());
            }
        }, context);

        RecyclerView iconRecycler = (RecyclerView) viewInflated.findViewById(R.id.recyclerView_icon);
        iconRecycler.setAdapter(iconPicker);
        iconRecycler.setHasFixedSize(true);

        LinearLayoutManager linearLayoutManagerCat = new LinearLayoutManager(context);
        linearLayoutManagerCat.setOrientation(LinearLayoutManager.HORIZONTAL);
        iconRecycler.setLayoutManager(linearLayoutManagerCat);

        builder.setView(viewInflated);
        builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                int lastAdapterPosition = iconPicker.getLastSelected(); // which icon was selected
                Category newCategory = new Category(input.getText().toString(), JetUUID.randomUUID(), lastAdapterPosition);
                onCategoryCreated.accept(newCategory);
                iconPicker.destroy();
                dialog.cancel();
            }
        });
        builder.setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                iconPicker.destroy();
                dialog.cancel();
            }
        });
        builder.show();
    }
}

File: app/src/main/java/com/sardox/timestamper/dialogs/AddCategoryDialog.kt
package com.sardox.timestamper.dialogs

import android.annotation.SuppressLint
import android.content.Context
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.widget.EditText

import com.sardox.timestamper.R
import com.sardox.timestamper.objects.Category
import com.sardox.timestamper.recyclerview.IconAdapter
import com.sardox.timestamper.types.JetUUID
import com.sardox.timestamper.utils.Consumer
import com.sardox.timestamper.utils.TimestampIcon

class AddCategoryDialog(context: Context, icons: List<TimestampIcon>, onCategoryCreated: (category: Category) -> Unit) {

    init {
        val builder = AlertDialog.Builder(context)
        @SuppressLint("InflateParams") val viewInflated = LayoutInflater.from(context).inflate(R.layout.new_category, null, false)
        val input = viewInflated.findViewById<EditText>(R.id.input_cat)
        val iconPicker = IconAdapter(icons, Consumer { icon -> input.setText(icon.description) }, context)

        val iconRecycler = viewInflated.findViewById<RecyclerView>(R.id.recyclerView_icon)
        iconRecycler.adapter = iconPicker
        iconRecycler.setHasFixedSize(true)

        val linearLayoutManagerCat = LinearLayoutManager(context)
        linearLayoutManagerCat.orientation = LinearLayoutManager.HORIZONTAL
        iconRecycler.layoutManager = linearLayoutManagerCat

        builder.setView(viewInflated)
        builder.setPositiveButton(R.string.save) { dialog, _ ->
            val lastAdapterPosition = iconPicker.lastSelected // which icon was selected
            val newCategory = Category(input.text.toString(), JetUUID.randomUUID(), lastAdapterPosition)
            onCategoryCreated.invoke(newCategory)
            iconPicker.destroy()
            dialog.cancel()
        }
        builder.setNegativeButton(android.R.string.cancel) { dialog, _ ->
            iconPicker.destroy()
            dialog.cancel()
        }
        builder.show()
    }
}